lib/src/binding_mqtt/mqtt_client.dart

```diff
@@ -0,0 +1,245 @@
+// Copyright 2022 The NAMIB Project Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+import 'dart:async';
+
+import 'package:mqtt_client/mqtt_client.dart';
+import 'package:mqtt_client/mqtt_server_client.dart';
+import 'package:typed_data/typed_buffers.dart';
+
+import '../core/content.dart';
+import '../core/credentials/basic_credentials.dart';
+import '../core/protocol_interfaces/protocol_client.dart';
+import '../core/security_provider.dart';
+import '../definitions/form.dart';
+import '../scripting_api/subscription.dart' as scripting_api;
+import 'mqtt_binding_exception.dart';
+import 'mqtt_config.dart';
+import 'mqtt_extensions.dart';
+import 'mqtt_subscription.dart';
+
+/// [ProtocolClient] for supporting the MQTT protocol.
+///
+/// Currently, only MQTT version 3.1.1 is supported.
+class MqttClient extends ProtocolClient {
+  /// Constructor.
+  MqttClient(
+    this._clientSecurityProvider,
+    MqttConfig? mqttConfig,
+  ) : _mqttConfig = mqttConfig ?? MqttConfig();
+
+  final MqttConfig _mqttConfig;
+
+  final ClientSecurityProvider? _clientSecurityProvider;
+
+  Future<BasicCredentials?> _obtainCredentials(
+    Uri uri,
+    Form? form, [
+    BasicCredentials? invalidCredentials,
+    bool unauthorized = false,
+  ]) async {
+    final requiresBasicAuthentication =
+        form?.requiresBasicAuthencation(invalidCredentials) ?? false;
+    final isDiscovery = form == null && unauthorized;
+
+    if (!(requiresBasicAuthentication || isDiscovery)) {
+      return null;
+    }
+
+    final basicCredentials = _clientSecurityProvider?.basicCredentialsCallback
+        ?.call(uri, form, invalidCredentials);
+
+    if (basicCredentials != null) {
+      return basicCredentials;
+    }
+
+    if (form != null) {
+      throw MqttBindingException(
+        'Form requires basic authencation but no credentials were provided.',
+      );
+    }
+
+    throw MqttBindingException(
+      'Discovery requires basic authencation but no credentials were provided.',
+    );
+  }
+
+  Future<MqttServerClient> _connectWithForm(Form form) async =>
+      _connect(form.resolvedHref, form);
+
+  Future<MqttServerClient> _connect(Uri brokerUri, Form? form) async {
+    final client = brokerUri.createClient();
+    final credentials = await _obtainCredentials(brokerUri, form);
+
+    MqttClientConnectionStatus? status;
```

Comment:
As this is a public field, it should be documented.

Reply:
Hmm, it is a local variable, though. Should it still be documented?

Reply:
Oh, I totally missed that, sorry, I thought it was a field in the class. Local variables don't need to be documented, no.